byterun/ints.c

```diff
@@ -828,3 +849,308 @@ CAMLprim value caml_nativeint_of_string(value s)
 {
   return caml_copy_nativeint(parse_intnat(s, 8 * sizeof(value), INTNAT_ERRMSG));
 }
+
+static int uint32_cmp(value v1, value v2)
+{
+  uint32_t u1 = Uint32_val(v1);
+  uint32_t u2 = Uint32_val(v2);
+  return (u1 > u2) - (u1 < u2);
+}
+
+static intnat uint32_hash(value v)
+{
+  return Uint32_val(v);
+}
+
+static void uint32_serialize(value v, uintnat *wsize_32, uintnat *wsize_64)
+{
+  caml_serialize_int_4(Uint32_val(v));
+  *wsize_32 = *wsize_64 = 4;
+}
+
+static uintnat uint32_deserialize(void *dst)
+{
+  *(uint32_t *)dst = caml_deserialize_uint_4();
+  return 4;
+}
+
+CAMLexport struct custom_operations caml_uint32_ops = {
+  "_u", 
+  custom_finalize_default,
+  uint32_cmp,
+  uint32_hash,
+  uint32_serialize,
+  uint32_deserialize, 
+  custom_compare_ext_default
+};
+
+CAMLprim value caml_copy_uint32(uint32_t u)
+{
+  value res = caml_alloc_custom(&caml_uint32_ops, 4, 0, 1);
+  Uint32_val(res) = u;
+  return res;
+}
+
+CAMLprim value caml_uint32_add(value a, value b)
+{ return caml_copy_uint32(Uint32_val(a) + Uint32_val(b)); }
+
+CAMLprim value caml_uint32_sub(value a, value b)
+{ return caml_copy_uint32(Uint32_val(a) - Uint32_val(b)); }
+
+CAMLprim value caml_uint32_mul(value a, value b)
+{ return caml_copy_uint32(Uint32_val(a) * Uint32_val(b)); }
+
+CAMLprim value caml_uint32_div(value a, value b)
+{ uint32_t dividend = Uint32_val(a);
+  uint32_t divisor = Uint32_val(b);
+  if (divisor == 0) caml_raise_zero_divide();
+  return caml_copy_uint32(dividend / divisor);
+}
```

Comment:
I personally have the impression that all of this would be easier to review and maintain if each `uint{32,64}` operation was right after the corresponding `int{32,64}` operation, which would also allow to share the ifdef logic between both (for instance `ARCH_ALIGN_INT64`). But the code looks ok.